### DISCUSSION TITLE
Handling errors from middleware from_fn

### DISCUSSION THREAD

**Discussion by zkrzyzanowski** [tokio-rs/axum#1151] [Q&A] [answered]
Hi!
First off, thanks for this crate! This is my first experience with building a web app with rust and axum has been great so far!

I've got a a middleware that I'm using to protect a route from being accessed if the user making the request isn't authorized. I'm new to rust, so I'm using the `middleware::from_fn` function to just convert a function to middleware. This function returns a `Result`, so I want to be able to log errors if they occur. I also only want the `/calculate` route to be protected by that middleware, but want all routes to have a timeout of 20s.

```
 let app = Router::new()
        .route("/calculate", post(calculate_route::run))
        .route_layer(middleware::from_fn(move |req, next| {
            routes::protected(
                req,
                next,
                some_arg,
            )
        }))
        .route("/status", get(|| status()))
        .layer(
            ServiceBuilder::new()
                // this middleware goes above `TimeoutLayer` because it will receive
                // errors returned by `TimeoutLayer`
                .layer(HandleErrorLayer::new(|_: BoxError| async {
                    StatusCode::REQUEST_TIMEOUT
                }))
                .layer(TimeoutLayer::new(Duration::from_secs(20))),
        );
```

* Is this the correct ordering of `layer`s so all have a 20s timeout but only `/cacluate` is protected by the `protected` middleware? 
* How can I take the result returned from `routes::protected` and log an error?

I tried this:
```
 .route_layer(
            ServiceBuilder::new()
                .layer(HandleErrorLayer::new(|_: BoxError| async {
                    error!("some error");
                    StatusCode::INTERNAL_SERVER_ERROR
                }))
                .layer(middleware::from_fn(move |req, next| {
                    firebase::protected(
                        req,
                        next,
                        maybe_emulator_host.clone(),
                        firebase_project_id.clone(),
                    )
                })),
        )
```
but get the following compiler error and I'm not quite sure what to do
```
error[E0277]: the trait bound `HandleError<axum::middleware::FromFn<[closure@src/main.rs:44:44: 51:18], Route<_>>, [closure@src/main.rs:41:46: 43:18], _>: Service<Request<_>>` is not satisfied
   --> src/main.rs:39:10
    |
39  |         .route_layer(
    |          ^^^^^^^^^^^ the trait `Service<Request<_>>` is not implemented for `HandleError<axum::middleware::FromFn<[closure@src/main.rs:44:44: 51:18], Route<_>>, [closure@src/main.rs:41:46: 43:18], _>`
    |
    = help: the following implementations were found:
              <HandleError<S, F, ()> as Service<Request<ReqBody>>>
              <HandleError<S, F, (T1, T2)> as Service<Request<ReqBody>>>
              <HandleError<S, F, (T1, T2, T3)> as Service<Request<ReqBody>>>
              <HandleError<S, F, (T1, T2, T3, T4)> as Service<Request<ReqBody>>>
            and 13 others
note: required by a bound in `Router::<B>::route_layer`
   --> /Users/zachkrzyzanowski/.cargo/registry/src/github.com-1ecc6299db9ec823/axum-0.5.9/src/routing/mod.rs:335:21
```

**Comment by davidpdrsn** [comment, marked as answer]
> Is this the correct ordering of layers so all have a 20s timeout but only /cacluate is protected by the protected middleware?

Yes, though I perfer `.route("/calculate", post(...).layer(...))`. I think thats more clear but it ultimately doesn't matter.

> How can I take the result returned from routes::protected and log an error?

I think you're misunderstanding how axum thinks about errors. Returning `500 Internal Server` (or some other status) from a handler is _not_ an error. Thats just another response. Errors only occur when you apply fallible middleware like [`TimeoutLayer`](https://docs.rs/tower/latest/tower/timeout/struct.TimeoutLayer.html). Notice that [it's error type](https://docs.rs/tower/latest/tower/timeout/struct.Timeout.html#impl-Service%3CRequest%3E) isn't `Infallible` which axum requires.

I recommend you just log error directly in `routes::protected`. There you also have the full context to provide a rich error message. Otherwise you're essentially lossily converting the error into HTTP only to log it.

Specifically for timeouts I recommend https://docs.rs/tower-http/latest/tower_http/timeout/index.html as it doesn't require `HandleErrorLayer`.

**Reply by zkrzyzanowski** [reply]
Wow, thanks for the quick response! That explanation makes sense, I was hoping to be able to log errors in one place, but I think I see what you're saying about losing the context when returning the response.

Thanks again for a thorough explanation!

**Reply by davidpdrsn** [reply]
You can do [this](https://github.com/tokio-rs/axum/blob/main/examples/error-handling-and-dependency-injection/src/main.rs) and log the error in the `IntoResponse` impl.
